### DISCUSSION TITLE
Empty Span in Java Application

### DISCUSSION THREAD

**Discussion by Taragos** [open-telemetry/opentelemetry-java-instrumentation#5115] [Q&A] [answered]
Hello,

I am currently trying to integrate the Auto Instrumentation Javaagent into an existing application and ran into some problems.
The application is a Karaf-based application in which a gRPC-Service runs. When I add the auto instrumentation agent, it correctly generates telemetry data and I can view it in my jaeger backend.

I now tried to access current span from within a method of the grpc Service (that method is definetly instrumented as I can see it in the Jaeger UI) but all I get is an empty span.
I did add the Opentelemetry API as a dependency and I can start my application without any error being thrown.

Javaagent Version: 1.9.1
Opentelemetry API Version: 1.9.1
io.grpc Version: 1.43

How I try to access it:
```
        Span span = Span.current();
        LOG.info(span.toString());
```

What I get:
```
PropagatedSpan{ImmutableSpanContext{traceId=00000000000000000000000000000000, spanId=0000000000000000, traceFlags=00, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=false}}
```

Is there something obvious I am missing here?

**Comment by laurit** [comment, marked as answer]
A small sample app that reproduces the issue would be helpful. If you have disabled instrumentations with `-Dotel.instrumentation.common.default-enabled=false` or in some other way then I'd suggest you try again with all instrumentations enabled and see whether it changes anything. If creating a sample app is too much effort then I'd suggest you start by examining `io.grpc.Context.current()` in debugger does it have `otel-context` in `keyValueEntries` field?

**Reply by Taragos** [reply]
Replicating it would take me a bit longer, so I would only do that if there is no other way.

I did not set any flags to disable certain instrumentations, so I guess they are all activated.

Debugging the application I can definetly see the `otel-context` variable right before I call `Span.current()`.
It does have a key `opentelemetry-trace-span-key` which is filled with a valid traceId and spanId.

The parentSpanContext is the default with an id of 0000..., but it is the first call in the chain so I guess that is normal.

Additionally when I inspect: `io.opentelemetry.javaagent.shaded.io.opentelemetry.context.Context.current()` I can also see the filled span.

But the `io.opentelemetry.context.current()`-call within the `Span.current()` call defaults to `root()` which in turn only returns `{ }`

**Reply by laurit** [reply]
Instead of `Span.current()` try evaluating `io.opentelemetry.context.Context.current()`. The returned object should be of type `AgentContextWrapper`. Check whether `agentContext` in the returned context contains `opentelemetry-trace-span-key`.

**Reply by Taragos** [reply]
This command results in a ClassNotFoundException...

**Reply by Taragos** [reply]
Thanks to that Exception I found the problem.
As I was running the application in a Karaf-container, I had to prepare the opentelemetry-packages as bundles.  
I accidentaly split the api and context packages into two different bundles.

After providing both packages via one bundle I can now acces the current Span.
Thank you for your help!

**Comment by xsmrcek** [comment]
@Taragos I had the same problem, after some investigation I found out, that I used different version of opentelemetry-api and agent :D
This version have to match agent:
 `<dependency>
            <groupId>io.opentelemetry</groupId>
            <artifactId>opentelemetry-api</artifactId>
            <version>0.16.0</version>
</dependency>
<dependency>
            <groupId>io.opentelemetry</groupId>
            <artifactId>opentelemetry-extension-annotations</artifactId>
            <version>0.16.0</version>
</dependency>`
Any change you did the same mistake? :)

**Reply by Taragos** [reply]
I have for both the same dependency: 1.9.1
(Both pulled via Maven)

**Reply by laurit** [reply]
Using an old api jar would explain this. You can verify this by evaluating `Span.class.getProtectionDomain().getCodeSource()`, see if it is loaded from where you'd expect it to be loaded.

**Reply by Taragos** [reply]
Executing that leaves me with a string in the format of `jar://bundle://<id>`, properly because of Karaf.
Taking a look at the Karaf container I can see the bundles that are loaded and the api and context bundles are both in version 1.9.1, which is the same as the one the opentelemtry-javaagent prints at the startup.
